File: ruby/snakegame/lib/snakegame/models/snake.rb
module Snakegame
  module Models
    class Snake
      VALID_DIRECTIONS = %i(left right top bottom)
      GROWTH_STEP = 3
      attr_accessor :coordinates, :current_step, :board_size, :last_direction

      def initialize(initial_size, board_size)
        self.coordinates = Array.new
        self.current_step = 1
        self.board_size = board_size
        self.last_direction = :right
        grow_initial(initial_size)
      end

      def move!(direction)
        validate_direction!(direction)
        return if should_nullify_move?(direction)

        new_x = coordinates[-1].x
        new_y = coordinates[-1].y
        growth_x = new_x
        growth_y = new_y

        if direction == :left
          new_x -= 1
          growth_x = new_x - 1
        elsif direction == :right
          new_x += 1
          growth_x = new_x + 1
        elsif direction == :top
          new_y -= 1
          growth_y = new_y - 1
        elsif direction == :bottom
          new_y += 1
          growth_y = new_y + 1
        end

        coordinate = normalize(Coordinate.new(new_x, new_y))
        check_collission!(coordinate)
        coordinates << coordinate
        grow_snake!(growth_x, growth_y)
        coordinates.shift
        self.current_step += 1
        self.last_direction = direction

        puts "Step: #{current_step}, Coordinates: #{coordinates.map(&:to_s)}"
      end

      private def grow_snake!(growth_x, growth_y)
        return unless should_grow_snake?

        puts "Growing..."
        growth_coordinate = normalize(Coordinate.new(growth_x, growth_y))
        check_collission!(growth_coordinate)
        coordinates << growth_coordinate
        self.current_step = 0
      end

      private def check_collission!(coordinate)
        raise Snakegame::Errors::HitTailError if coordinates.include?(coordinate)
      end

      private def should_grow_snake?
        current_step == GROWTH_STEP
      end

      private def should_nullify_move?(direction)
        (direction == :left && last_direction == :right) ||
        (direction == :right && last_direction == :left) ||
        (direction == :top && last_direction == :bottom) ||
        (direction == :bottom && last_direction == :top)
      end

      private def validate_direction!(direction)
        unless VALID_DIRECTIONS.include?(direction)
          raise Snakegame::Errors::InvalidDirectionError.new("Invalid direction")
        end
      end

      private def normalize(coordinate)
        coordinate.x %= board_size if coordinate.x > board_size
        coordinate.x = board_size - coordinate.x if coordinate.x <= 0

        coordinate.y %= board_size if coordinate.y > board_size
        coordinate.y = board_size - coordinate.y if coordinate.y <= 0

        coordinate
      end

      private def grow_initial(initial_size)
        initial_size.times do |time|
          self.coordinates << Snakegame::Models::Coordinate.new(time + 1, 1)
        end
      end
    end
  end
end

File: java/snakegame/src/main/java/gagan93jtg/lld/models/Snake.java
package gagan93jtg.lld.models;

import gagan93jtg.lld.errors.HitTrailError;
import gagan93jtg.lld.errors.InvalidDirectionError;

import java.util.ArrayList;
import java.util.EnumSet;

enum ValidDirection {LEFT, RIGHT, BOTTOM, TOP};

public class Snake {
  public static final int GROWTH_STEP = 3;

  private int currentStep, boardSize;
  private ValidDirection lastDirection;
  private ArrayList<Coordinate> coordinates;

  public Snake(int initialSize, int boardSize) {
    this.coordinates = new ArrayList<>();
    this.currentStep = 1;
    this.boardSize = boardSize;
    this.lastDirection = ValidDirection.RIGHT;
    growInitial(initialSize);
  }

  ArrayList<Coordinate> getCoordinates() {
    return coordinates;
  }

  void setCoordinates(ArrayList<Coordinate> coordinates) {
    this.coordinates = coordinates;
  }

  void setLastDirection(ValidDirection lastDirection) {
    this.lastDirection = lastDirection;
  }

  public void move(ValidDirection direction) {
    if (moveIsInvalid(direction)) {
      return;
    }

    int newX, newY, growthX, growthY;
    newX = coordinates.get(coordinates.size() - 1).getX();
    newY = coordinates.get(coordinates.size() - 1).getY();
    growthX = newX;
    growthY = newY;

    if (direction == ValidDirection.LEFT) {
      newX -= 1;
      growthX -= 2;
    } else if (direction == ValidDirection.RIGHT) {
      newX += 1;
      growthX += 2;
    } else if (direction == ValidDirection.TOP) {
      newY -= 1;
      growthY -= 2;
    } else if (direction == ValidDirection.BOTTOM) {
      newY += 1;
      growthY += 2;
    }

    Coordinate coordinate = normalize(new Coordinate(newX, newY));
    checkCollission(coordinate);
    coordinates.add(coordinate);
    growSnake(growthX, growthY);
    coordinates.remove(0);
    this.currentStep += 1;
    this.lastDirection = direction;

    System.out.println("Step: " + currentStep + ", Coordinates: " + coordinates);
  }

  private void growSnake(int x, int y) {
    if (shouldNotGrowSnake()) {
      return;
    }

    System.out.println("Growing...");
    Coordinate coordinate = normalize(new Coordinate(x, y));
    checkCollission(coordinate);
    coordinates.add(coordinate);
    this.currentStep = 0;
  }

  private boolean shouldNotGrowSnake() {
    return currentStep != GROWTH_STEP;
  }

  private Coordinate normalize(Coordinate coordinate) {
    if (coordinate.getX() > boardSize)
      coordinate.setX(coordinate.getX() % boardSize);
    else if (coordinate.getX() <= 0)
      coordinate.setX(boardSize - coordinate.getX());

    if (coordinate.getY() > boardSize)
      coordinate.setY(coordinate.getY() % boardSize);
    else if (coordinate.getY() <= 0)
      coordinate.setY(boardSize - coordinate.getY());

    return coordinate;
  }

  private void checkCollission(Coordinate coordinate) {
    if (coordinates.contains(coordinate))
      throw new HitTrailError("Game over");
  }

  private boolean moveIsInvalid(ValidDirection direction) {
    return (direction == ValidDirection.LEFT && lastDirection == ValidDirection.RIGHT) ||
            (direction == ValidDirection.RIGHT && lastDirection == ValidDirection.LEFT) ||
            (direction == ValidDirection.TOP && lastDirection == ValidDirection.BOTTOM) ||
            (direction == ValidDirection.BOTTOM && lastDirection == ValidDirection.TOP);
  }

  private void growInitial(int initialSize) {
    for (int i = 0; i < initialSize; i++) {
      this.coordinates.add(new Coordinate(i + 1, 1));
    }
  }
}
